### ISSUE TITLE
i2c not working with re-enterancy changes

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#374]
**_Reported by Anas Nashif:_**

Just found out I2C is not working as expected with at least 2 of the samples, bisected and found this to be the culprit..
 
```
core:zephyr((no branch, bisect started on master)): git bisect good
bf45b0bd870b57433ef5b5ef62859302416d0ca8 is the first bad commit
commit bf45b0bd870b57433ef5b5ef62859302416d0ca8
Author: Sergio Rodriguez <sergio.sf.rodriguez@intel.com>
Date:   Tue May 10 16:28:11 2016 -0700
 
    i2c: Use locking mechanism to guard critical regions.
 
    This will allow the driver to be safe from fibers and task
 
    Change-Id: I6c4c4fc387cf334f0287b8a02982be1941a311fd
    Signed-off-by: Sergio Rodriguez <sergio.sf.rodriguez@intel.com>
 
:040000 040000 b9f4876f58c9959efbf4f31d23d19209275c52eb d5aa442cae837e79612b6be7c6faafa65351fc50 M                drivers


```

(Imported from Jira ZEP-419)

**Comment by nashif** [comment]
**_by Sergio Rodriguez:_**

Seems I found the issue the patch is in https://gerrit.zephyrproject.org/r/#/c/2322/, the sync data must be set before we enable the interrupts

**Comment by nashif** [comment]
**_by Anas Nashif:_**

what about SPI?

**Comment by nashif** [comment]
**_by Sergio Rodriguez:_**

Looking at it right now, it seems the same issue I will update a patch

**Comment by nashif** [comment]
**_by Lei Liu:_**

Fixed on Zephyr v1.4.0-rc2
